geoportal/c2cgeoportal_geoportal/views/login.py

```diff
@@ -0,0 +1,340 @@
+                                    "otp_uri": pyotp.TOTP(user.tech_data["2fa_totp_secret"]).provisioning_uri(
+                                        user.email, issuer_name=self.two_factor_issuer_name
+                                    ),
+                                }
+                            ),
+                            headers=(("Content-Type", "text/json"),),
+                        ),
+                    )
+                otp = self.request.POST.get("otp")
+                if otp is None:
+                    raise HTTPBadRequest("The second factor is missing.")
+                if not self._validate_2fa_totp(user, otp):
+                    LOG.info("The second factor is wrong for user '%s'.", user)
+                    raise HTTPUnauthorized("See server logs for details")
+            user.update_last_login()
+            user.tech_data["consecutive_failed"] = "0"
+
+            if not user.is_password_changed:
+                return set_common_headers(
+                    self.request,
+                    "login",
+                    NO_CACHE,
+                    response=Response(
+                        json.dumps(
+                            {
+                                "username": user.username,
+                                "is_password_changed": False,
+                                "two_factor_enable": True,
+                            }
+                        ),
+                        headers=(("Content-Type", "text/json"),),
+                    ),
+                )
+
+            headers = remember(self.request, username)
+            LOG.info("User '%s' logged in.", username)
+            came_from = self.request.params.get("came_from")
+            if came_from:
+                return HTTPFound(location=came_from, headers=headers)
+            else:
+                headers.append(("Content-Type", "text/json"))
+                return set_common_headers(
+                    self.request,
+                    "login",
+                    NO_CACHE,
+                    response=Response(
+                        json.dumps(self._user(self.request.get_user(username))), headers=headers
+                    ),
+                )
+        else:
+            user = models.DBSession.query(static.User).filter(static.User.username == login).one_or_none()
+            if user and not user.deactivated:
+                if "consecutive_failed" not in user.tech_data:
+                    user.tech_data["consecutive_failed"] = "0"
+                user.tech_data["consecutive_failed"] = str(int(user.tech_data["consecutive_failed"]) + 1)
+                if int(user.tech_data["consecutive_failed"]) >= self.request.registry.settings.get(
+                    "authentication", {}
+                ).get("max_consecutive_failures", sys.maxsize):
+                    user.deactivated = True
+                    user.tech_data["consecutive_failed"] = "0"
```

Comment:
Easy DOS, here. Maybe you want to disable that for admin?

Reply:
I think that we should know that someone try to get the admin password, and we will reset it directly in the database ...